archive/app/controllers/ArchiveController.scala

```diff
@@ -66,9 +68,31 @@ object ArchiveController extends Controller with Logging with ExecutionContexts
     case _ => false
   }
 
-  private def destinationFor(path: String) = DynamoDB.destinationFor(path).map(_.filterNot { destination =>
-      linksToItself(path, destination.location)
-  })
+  def retainShortUrlCampaign(path: String)(destination: Destination): Destination = {
+    // if the path is a short url with a campaign, and the destination doesn't have a campaign, pass it through the redirect.
+    val shortUrlWithCampaign = """.*www\.theguardian\.com/p/[\w\d]+/([\w\d]+)$""".r
+    val urlWithCampaignParam = """.*www\.theguardian\.com.*?.*CMP=.*$""".r
+
+    val destinationHasCampaign = destination.location match {
+      case shortUrlWithCampaign(_) => true
+      case urlWithCampaignParam() => true
+      case _ => false
+    }
+
+    path match {
+      case shortUrlWithCampaign(campaign) if !destinationHasCampaign => {
+        val uri = javax.ws.rs.core.UriBuilder.fromPath(destination.location)
```

Comment:
this library does actually build `java.net.URI` objects, they are fine. The thing I have bad history with is [`URLEncoder`](http://docs.oracle.com/javase/7/docs/api/java/net/URLEncoder.html), which looks like it can build sensible `java.net.URI` objects, but is specifically for encoding HTML forms urls (`application/x-www-form-urlencoded`). Despite its name, `URLEncoder` isn't actually rfc2396 compliant. This has bitten us regularly, so I recently added this Jersey library to help with the issue. I explicitly do not condone url building, it's well dodgy.